politeiawww/invoices.go

```diff
@@ -0,0 +1,447 @@
+	BackendInvoiceMDChangeVersion = 1
+)
+
+// processNewInvoice tries to submit a new proposal to politeiad.
+func (p *politeiawww) processNewInvoice(ni cms.NewInvoice, u *user.User) (*cms.NewInvoiceReply, error) {
+	log.Tracef("processNewInvoice")
+
+	err := validateInvoice(ni, u)
+	if err != nil {
+		return nil, err
+	}
+
+	name := strconv.Itoa(int(ni.Year)) + strconv.Itoa(int(ni.Month)) + u.Username
+
+	md, err := encodeBackendInvoiceMetadata(BackendInvoiceMetadata{
+		Version:   BackendInvoiceMetadataVersion,
+		Timestamp: time.Now().Unix(),
+		Name:      name,
+		PublicKey: ni.PublicKey,
+		Signature: ni.Signature,
+		Month:     ni.Month,
+		Year:      ni.Year,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Setup politeiad request
+	challenge, err := util.Random(pd.ChallengeSize)
+	if err != nil {
+		return nil, err
+	}
+	n := pd.NewRecord{
+		Challenge: hex.EncodeToString(challenge),
+		Metadata: []pd.MetadataStream{{
+			ID:      mdStreamGeneral,
+			Payload: string(md),
+		}},
+		Files: convertPropFilesFromWWW(ni.Files),
+	}
+
+	// Handle test case
+	if p.test {
+		tokenBytes, err := util.Random(pd.TokenSize)
+		if err != nil {
+			return nil, err
+		}
+
+		testReply := pd.NewRecordReply{
+			CensorshipRecord: pd.CensorshipRecord{
+				Token: hex.EncodeToString(tokenBytes),
+			},
+		}
+
+		return &cms.NewInvoiceReply{
+			CensorshipRecord: convertPropCensorFromPD(testReply.CensorshipRecord),
+		}, nil
+	}
+
+	// Send the newrecord politeiad request
+	responseBody, err := p.makeRequest(http.MethodPost,
+		pd.NewRecordRoute, n)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Infof("Submitted invoice name: %v", name)
+	for k, f := range n.Files {
+		log.Infof("%02v: %v %v", k, f.Name, f.Digest)
+	}
+
+	// Handle newRecord response
+	var pdReply pd.NewRecordReply
+	err = json.Unmarshal(responseBody, &pdReply)
+	if err != nil {
+		return nil, fmt.Errorf("Unmarshal NewInvoiceReply: %v", err)
+	}
+
+	// Verify NewRecord challenge
+	err = util.VerifyChallenge(p.cfg.Identity, challenge, pdReply.Response)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create change record since we are immediately considering this Vetted and
+	// New Invoice Status.
+	changes := BackendInvoiceMDChange{
+		Version:   BackendInvoiceMDChangeVersion,
+		Timestamp: time.Now().Unix(),
+		NewStatus: cms.InvoiceStatusNew,
+	}
+
+	var pdSetUnvettedStatusReply pd.SetUnvettedStatusReply
+	challenge, err = util.Random(pd.ChallengeSize)
+	if err != nil {
+		return nil, err
+	}
+
+	blob, err := json.Marshal(changes)
+	if err != nil {
+		return nil, err
+	}
+
+	sus := pd.SetUnvettedStatus{
+		Token:     pdReply.CensorshipRecord.Token,
+		Status:    pd.RecordStatusPublic,
+		Challenge: hex.EncodeToString(challenge),
+		MDAppend: []pd.MetadataStream{
+			{
+				ID:      mdStreamChanges,
+				Payload: string(blob),
+			},
+		},
+	}
+
+	// Send SetUnvettedStatus request to politeiad
+	responseBody, err = p.makeRequest(http.MethodPost, pd.SetUnvettedStatusRoute, sus)
```

Comment:
80 column limit.  

Different repos have different philosophies on the 80 column limit.  Some treat it as a hard limit, some treat it as a soft limit.  We tend to treat it as a hard limit throughout the politeia repo.  The exceptions being function declarations and in-line comments of struct fields.